### PR TITLE
Refactoring with the help of Sourcery

### PR DESCRIPTION
**Pull request by croghostrider**
## Description

I do code refactoring with the help of Sourcery.
```
Issues by Rule ID
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Rule ID                              ┃ Count ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│ assign-if-exp                        │    26 │
│ raise-from-previous-error            │    25 │
│ swap-if-expression                   │    23 │
│ reintroduce-else                     │    22 │
│ use-fstring-for-concatenation        │    13 │
│ merge-nested-ifs                     │    11 │
│ remove-unnecessary-cast              │    11 │
│ boolean-if-exp-identity              │     6 │
│ use-next                             │     6 │
│ list-comprehension                   │     5 │
│ use-any                              │     5 │
│ merge-else-if-into-elif              │     4 │
│ remove-redundant-if                  │     4 │
│ remove-redundant-slice-index         │     3 │
│ swap-nested-ifs                      │     3 │
│ hoist-similar-statement-from-if      │     2 │
│ move-assign-in-block                 │     2 │
│ remove-zero-from-range               │     2 │
│ replace-interpolation-with-fstring   │     2 │
│ aug-assign                           │     1 │
│ de-morgan                            │     1 │
│ inline-immediately-returned-variable │     1 │
│ introduce-default-else               │     1 │
│ invert-any-all                       │     1 │
│ merge-duplicate-blocks               │     1 │
│ min-max-identity                     │     1 │
│ simplify-len-comparison              │     1 │
│ switch                               │     1 │
├──────────────────────────────────────┼───────┤
│ Total                                │   187 │
```
The refactorings don't change the code, should I add something to the changelog?

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] The documentation has been adjusted accordingly
- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] The changes are documented in the changelog (docs/changelog.md)

### CODE REVIEW

**Comment by farmio** [comment]
Hi 👋! 
While most changes from your first commit seem reasonable at a first glance, looking at the changes form [a2f830e](https://github.com/XKNX/xknx/pull/1157/commits/a2f830ea5806b12b3e47fd52b92f80e355d00d6f) I don't really see a point in any of these. It's mostly redundant type casting 🤨

**Comment by croghostrider** [comment]
hi @farmio, I agree with you, but something has broken the code from the "remove-unnecessary-cast" rule.
Now is the pull request smaller and easier to review.
It should pass the testing now.

Here the fixes:
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Rule ID                              ┃ Count ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│ swap-if-expression                   │    18 │
│ use-fstring-for-concatenation        │    13 │
│ merge-nested-ifs                     │    10 │
│ use-next                             │     6 │
│ list-comprehension                   │     5 │
│ use-any                              │     5 │
│ merge-else-if-into-elif              │     4 │
│ remove-redundant-if                  │     4 │
│ remove-redundant-slice-index         │     3 │
│ remove-zero-from-range               │     2 │
│ replace-interpolation-with-fstring   │     2 │
│ aug-assign                           │     1 │
│ de-morgan                            │     1 │
│ inline-immediately-returned-variable │     1 │
│ invert-any-all                       │     1 │
│ merge-duplicate-blocks               │     1 │
│ move-assign-in-block                 │     1 │
│ simplify-len-comparison              │     1 │
│ switch                               │     1 │
├──────────────────────────────────────┼───────┤
│ Total                                │    80 │
└──────────────────────────────────────┴───────┘
```

**Comment by croghostrider** [comment]
Hi @farmio 
What is the reason that the pull request is not merged?

**Comment by farmio** [comment]
I did not find any time to review the changes.

File: xknx/core/state_updater.py [hunk 1]
```diff
@@ -209,9 +209,8 @@ async def connection_state_change_callback(
         if state == XknxConnectionState.CONNECTED:
             if not self.started:
                 self._start()
-        else:
-            if self.started:
-                self._stop()
```
Comment (farmio) [inline xknx/core/state_updater.py:214]: 🧑‍💻

File: xknx/devices/climate_mode.py [hunk 2]
```diff
@@ -308,11 +308,10 @@ async def process_group_write(self, telegram: "Telegram") -> None:
         """Process incoming and outgoing GROUP WRITE telegram."""
         if self.supports_operation_mode:
             for rv_mode in self._iter_operation_remote_values():
-                if await rv_mode.process(telegram):
-                    #  ignore inactive RemoteValueBinaryOperationMode
-                    if rv_mode.value:
-                        await self._set_internal_operation_mode(rv_mode.value)
-                        return
```
Comment (farmio) [inline xknx/devices/climate_mode.py:315]: 🧑‍💻

File: xknx/devices/cover.py [hunk 3]
```diff
@@ -355,9 +355,10 @@ async def process_group_write(self, telegram: "Telegram") -> None:
                     target_position=self.travelcalculator.position_closed
                 )
         # stop from bus
-        if await self.stop_.process(telegram) or await self.step.process(telegram):
-            if self.is_traveling():
-                await self._stop_position_update()
```
Comment (farmio) [inline xknx/devices/cover.py:360]: 🧑‍💻

File: xknx/devices/switch.py [hunk 4]
```diff
@@ -86,13 +86,16 @@ async def set_off(self) -> None:
 
     async def process_group_write(self, telegram: "Telegram") -> None:
         """Process incoming and outgoing GROUP WRITE telegram."""
-        if await self.switch.process(telegram):
-            if self.reset_after is not None and self.switch.value:
-                self._reset_task = self.xknx.task_registry.register(
-                    name=self._reset_task_name,
-                    async_func=partial(self._reset_state, self.reset_after),
-                    track_task=True,
-                ).start()
```
Comment (farmio) [inline xknx/devices/switch.py:95]: 🧑‍💻

File: xknx/dpt/dpt.py [hunk 5]
```diff
@@ -109,20 +109,28 @@ def transcoder_by_dpt(
         cls: T, dpt_main: int, dpt_sub: int | None = None
     ) -> T | None:
         """Return Class reference of DPTBase subclass with matching DPT number."""
-        for dpt in cls.dpt_class_tree():
-            if dpt.has_distinct_dpt_numbers():
-                if dpt_main == dpt.dpt_main_number and dpt_sub == dpt.dpt_sub_number:
-                    return dpt
-        return None
```
Comment (farmio) [inline xknx/dpt/dpt.py:116]: 🧑‍💻

File: xknx/dpt/dpt.py [hunk 6]
```diff
@@ -109,20 +109,28 @@ def transcoder_by_dpt(
...
+        )
 
     @classmethod
     def transcoder_by_value_type(cls: T, value_type: str) -> T | None:
         """Return Class reference of DPTBase subclass with matching value_type."""
-        for dpt in cls.dpt_class_tree():
-            if dpt.has_distinct_value_type():
-                if value_type == dpt.value_type:
-                    return dpt
-        return None
```
Comment (farmio) [inline xknx/dpt/dpt.py:125]: 🧑‍💻

File: xknx/dpt/dpt_time.py [hunk 7]
```diff
@@ -53,13 +53,15 @@ def to_knx(cls, value: time.struct_time) -> tuple[int, int, int]:
             )
 
         _default_time = time.strptime("", "")
-        weekday = 0
         # if 0 year, 1 month, 2 day, 6 weekday, 7 yearday, 8 dst are equal to default assume "any weekday" (0)
-        for index in [0, 1, 2, 6, 7, 8]:
-            if value[index] is not _default_time[index]:
-                weekday = value.tm_wday + 1
-                break
-
```
Comment (farmio) [inline xknx/dpt/dpt_time.py:62]: 🧑‍💻

File: xknx/telegram/telegram.py [hunk 8]
```diff
@@ -52,15 +52,17 @@ def __init__(
...
         self.tpci: TPCI
         if tpci is None:
-            if isinstance(destination_address, GroupAddress):
-                if destination_address.raw == 0:
-                    self.tpci = TDataBroadcast()
-                else:
-                    self.tpci = TDataGroup()
-            elif isinstance(destination_address, IndividualAddress):
-                self.tpci = TDataIndividual()
-            else:  # InternalGroupAddress
```
Comment (farmio) [inline xknx/telegram/telegram.py:62]: 🧑‍💻
This is sorted by the most common use case - so first GroupAddress is handled. Lastly InternalGroupAddress as this is hardly used anyway and more of an edge-case.
Other than that we could firstly check for `IndividualAddress`, then for `destination_address.raw == 0` and assume all other are `TDataGroup()`.

File: xknx/secure/keyring.py [hunk 9]
```diff
@@ -258,11 +258,14 @@ def __init__(self) -> None:
 
     def get_device_by_interface(self, interface: XMLInterface) -> XMLDevice | None:
         """Get the device for a given interface."""
-        for device in self.devices:
-            if device.individual_address == interface.host:
-                return device
-
-        return None
```
Comment (farmio) [inline xknx/secure/keyring.py:265]: 🧑‍💻

**Review by farmio — CHANGES_REQUESTED** [review]
Thank you for your contirbution! 👍

I did mark some changes with "🧑‍💻". Here I think the original version is better readable / understandable and would prefer to keep it as is.

File: xknx/core/state_updater.py (hunk 1 above)
Comment (croghostrider) [reply xknx/core/state_updater.py:214]: I don't entirely agree here, but if you want I can change it back to the original.

Comment (farmio) [reply xknx/core/state_updater.py:214]: Nah, that's ok. I wasn't sure about that one either.

**Review by farmio — APPROVED** [review]
Thank you very much! 😃

**Pull request merged**